sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -223,47 +205,13 @@ object AnsiTypeCoercion extends TypeCoercionBase {
   override def canCast(from: DataType, to: DataType): Boolean = Cast.canAnsiCast(from, to)
 
   object PromoteStrings extends TypeCoercionRule {
-    private def castExpr(expr: Expression, targetType: DataType): Expression = {
-      expr.dataType match {
-        case NullType => Literal.create(null, targetType)
-        case l if l != targetType => Cast(expr, targetType)
-        case _ => expr
-      }
-    }
 
     override def transform: PartialFunction[Expression, Expression] = {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
-
-      case b @ BinaryOperator(left, right)
-        if findWiderTypeForString(left.dataType, right.dataType).isDefined =>
-        val promoteType = findWiderTypeForString(left.dataType, right.dataType).get
-        b.withNewChildren(Seq(castExpr(left, promoteType), castExpr(right, promoteType)))
-
-      case Abs(e @ StringTypeExpression(), failOnError) => Abs(Cast(e, DoubleType), failOnError)
-      case m @ UnaryMinus(e @ StringTypeExpression(), _) =>
-        m.withNewChildren(Seq(Cast(e, DoubleType)))
-      case UnaryPositive(e @ StringTypeExpression()) => UnaryPositive(Cast(e, DoubleType))
-
-      case d @ DateAdd(left @ StringTypeExpression(), _) =>
-        d.copy(startDate = Cast(d.startDate, DateType))
-      case d @ DateAdd(_, right @ StringTypeExpression()) =>
-        d.copy(days = Cast(right, IntegerType))
-      case d @ DateSub(left @ StringTypeExpression(), _) =>
-        d.copy(startDate = Cast(d.startDate, DateType))
-      case d @ DateSub(_, right @ StringTypeExpression()) =>
-        d.copy(days = Cast(right, IntegerType))
-
-      case s @ SubtractDates(left @ StringTypeExpression(), _, _) =>
-        s.copy(left = Cast(s.left, DateType))
-      case s @ SubtractDates(_, right @ StringTypeExpression(), _) =>
-        s.copy(right = Cast(s.right, DateType))
-      case t @ TimeAdd(left @ StringTypeExpression(), _, _) =>
-        t.copy(start = Cast(t.start, TimestampType))
-      case t @ SubtractTimestamps(left @ StringTypeExpression(), _, _, _) =>
-        t.copy(left = Cast(t.left, t.right.dataType))
-      case t @ SubtractTimestamps(_, right @ StringTypeExpression(), _, _) =>
-        t.copy(right = Cast(right, t.left.dataType))
+      case withChildrenResolved
+          if AnsiStringPromotionTypeCoercion.apply.isDefinedAt(withChildrenResolved) =>
```

Comment:
why do we use `PartialFunction`? I think it's simpler if we just use `Expression => Expression`, then the code here can be `case withChildrenResolved => AnsiStringPromotionTypeCoercion(withChildrenResolved)`. `AnsiStringPromotionTypeCoercion` should leave the input expression unchanged if the pattern match does not hit.

Reply:
So I used `PartialFunction` in order to be able to utilize `isDefinedAt`. If we do `Expression => Expression` I think we would have to add default cases for all these functions in order to avoid match errors. I didn't want to introduce that change as I think there definitely are parts of the code that don't expect `identity` when they don't match against anything but perform some custom logic

Reply:
I think Mihailo is right. We preserve the exact behavior as it was here before our refactoring. It's not correct to introduce some behavior changes in some giant refactoring PR

Reply:
The current code does the pattern match twice: one with `isDefinedAt` and one with `apply`. How about
```
case withChildrenResolved => AnsiStringPromotionTypeCoercion.applyOrElse(withChildrenResolved, other => other)
```

Reply:
But doesn't this run into the same issue where `case withChildrenResolved` is always matched because of `orElse`? Point of using `isDefinedAt` is to not match this case at all, but keep the current behavior as is

Reply:
The`transform` is used in the base trait `TypeCoercionRule` and `CombinedTypeCoercionRule`, and no one calls its `isDefinedAt` function.
```
          withPropagatedTypes.transformExpressionsUpWithPruning(
            AlwaysProcess.fn, ruleId)(typeCoercionFn)
...
    override def transform: PartialFunction[Expression, Expression] = {
      val transforms = rules.map(_.transform)
      Function.unlift { e: Expression =>
        val result = transforms.foldLeft(e) {
          case (current, transform) => transform.applyOrElse(current, identity[Expression])
        }
        if (result ne e) {
          Some(result)
        } else {
          None
        }
      }
    }
```


In the meanwhile, doing pattern match twice might cause performance regressions.

Reply:
Changed all `PartialFunction` to `Expression => Expression` functions